File: Java/src/com/liu/LeetCode/First/exercise/LC057_InsertInterval.java
package com.liu.LeetCode.First.exercise;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by liu on 2017/9/30.
 */
public class LC057_InsertInterval {
    /**
     * Given a set of non-overlapping intervals, insert a new interval into the intervals (merge if necessary).
     * <p>
     * You may assume that the intervals were initially sorted according to their start times.
     * <p>
     * Example 1:
     * Given intervals [1,3],[6,9], insert and merge [2,5] in as [1,5],[6,9].
     * <p>
     * Example 2:
     * Given [1,2],[3,5],[6,7],[8,10],[12,16], insert and merge [4,9] in as [1,2],[3,10],[12,16].
     * <p>
     * This is because the new interval [4,9] overlaps with [3,5],[6,7],[8,10].
     */

    public static class Interval {
        public int start;
        public int end;

        public Interval(int s, int e) {
            start = s;
            end = e;
        }
    }

    public static List<Interval> getInsertIntervalMy1(List<Interval> intervals, Interval newInterval) {
        List<Interval> result = new ArrayList<>();
        if (null != newInterval) {
            // 先加入需要合并的数据
            intervals.add(newInterval);
        }
        int length = intervals.size();
        if (length <= 1) {
            return intervals;
        }
        // 先进行排序,这种lambda表达式只有Java8以上才能支持
        intervals.sort((i1, i2) -> Integer.compare(i1.start, i2.start));
        result.add(intervals.get(0));
        for (int i = 1; i < length; i++) {
            int resultLength = result.size();
            Interval lastInterval = result.get(resultLength - 1);
            Interval interval = intervals.get(i);
            if (lastInterval.end >= interval.start) {
                result.remove(resultLength - 1);
                // 用于规避[1,4][2,3]
                int min = Math.min(lastInterval.start, interval.start);
                int max = Math.max(lastInterval.end, interval.end);
                result.add(new Interval(min, max));
            } else {
                result.add(interval);
            }
        }
        return result;
    }

    public static List<Interval> getInsertIntervalAnswer1(List<Interval> intervals, Interval newInterval) {
        // 也是一种很直白的方案
        List<Interval> result = new ArrayList<>();
        for (Interval i : intervals) {
            if (newInterval == null || i.end < newInterval.start)
                result.add(i);
            else if (i.start > newInterval.end) {
                result.add(newInterval);
                result.add(i);
                newInterval = null;
            } else {
                newInterval.start = Math.min(newInterval.start, i.start);
                newInterval.end = Math.max(newInterval.end, i.end);
            }
        }
        if (newInterval != null)
            result.add(newInterval);
        return result;
    }
}

File: Python/LeetCodeTest/lc057_InsertInterval.py
from typing import List

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]: # 先依照每个段落的开始值排序，然后只需要比较结束值
        all_data = intervals.copy()
        all_data.append(newInterval)
        all_data.sort(key=lambda x: x[0])
        output = []
        for interval in all_data:
            if len(output) == 0 or output[-1][1] < interval[0]:
                output.append(interval)
            else:
                output[-1][1] = max(output[-1][1], interval[1])
        return output

    def insert_error(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]: # 无法处理类似[[1,2],[3,4]],[0,5]这种新数据完全覆盖全部老数据的情况
        if len(intervals) == 0:
            return [newInterval]
        found_start = False
        found_end = False
        start = 0
        end = 0
        output = []
        for i in range(len(intervals)):
            interval = intervals[i]
            if interval[1] >= newInterval[0] and interval[0] <= newInterval[0]:
                if not found_start:
                    start = i
                    found_start = True
            elif interval[1] >= newInterval[1] and interval[0] <= newInterval[1]:
                if not found_end:
                    end = i
                    found_end = True
            elif interval[0] >= newInterval[0] and interval[1] <= newInterval[1]:
                if not found_end and not found_start:
                    found_start = True
                    found_end = True
                    start = end = i
        if found_start and not found_end:
            end = start
            found_end = True
        if found_end and not found_start:
            end = start
            found_start = True
        if found_end and found_start:
            cache = []
            found = False
            for i in range(len(intervals)):
                interval = intervals[i]
                if i == start:
                    found = True
                    cache.append(min(interval[0], newInterval[0]))
                if not found:
                    output.append(interval)
                if i == end:
                    found = False
                    cache.append(max(interval[1], newInterval[1]))
                    output.append(cache)
        else:
            output = intervals.copy()
            output.append(newInterval)
            output.sort(key=lambda x: x[0])
        return output

if __name__ == "__main__":
    solution = Solution()
    result = solution.insert([[1, 5]], [0, 3])
    print(result)
    result = solution.insert([[1, 5], [6, 8]], [0, 9])
    print(result)
    result = solution.insert([[1, 5]], [6, 8])
    print(result)
    result = solution.insert([[1, 3], [6, 9]], [2, 5])
    print(result)
    result = solution.insert([[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]], [4, 8])
    print(result)
